crates/driver/src/infra/blockchain/gas.rs

```diff
@@ -0,0 +1,75 @@
+/// Wrapper around the gas estimation library.
+/// Also llows to add additional tip to the gas price. This is used to
+/// increase the chance of a transaction being included in a block, in case
+/// private submission networks are used.
+use {
+    super::Error,
+    crate::{domain::eth, infra::mempool},
+    ethcontract::dyns::DynWeb3,
+    gas_estimation::{nativegasestimator::NativeGasEstimator, GasPriceEstimating},
+};
+
+type MaxAdditionalTip = f64;
+type AdditionalTipPercentage = f64;
+type AdditionalTip = (MaxAdditionalTip, AdditionalTipPercentage);
+
+pub struct GasPriceEstimator {
+    gas: NativeGasEstimator,
+    additional_tip: Option<AdditionalTip>,
+}
+
+impl GasPriceEstimator {
+    pub async fn new(web3: &DynWeb3, mempools: &[mempool::Config]) -> Result<Self, Error> {
+        let gas = NativeGasEstimator::new(web3.transport().clone(), None)
+            .await
+            .map_err(Error::Gas)?;
+        let additional_tip = mempools
+            .iter()
+            .find(|mempool| matches!(mempool.kind, mempool::Kind::Flashbots { .. }))
+            .map(|mempool| {
+                (
+                    match mempool.kind {
+                        mempool::Kind::Flashbots {
+                            max_additional_tip, ..
+                        } => max_additional_tip,
+                        _ => unreachable!(),
+                    },
+                    mempool.additional_tip_percentage,
+                )
+            });
+        Ok(Self {
+            gas,
+            additional_tip,
+        })
+    }
+
+    /// Estimates the gas price for a transaction.
+    /// If additional tip is configured, it will be added to the gas price. This
+    /// is to increase the chance of a transaction being included in a block, in
+    /// case private submission networks are used.
+    pub async fn estimate(&self) -> Result<eth::GasPrice, Error> {
+        self.gas
+            .estimate()
+            .await
+            .map(|mut estimate| {
+                let estimate = match self.additional_tip {
+                    Some((max_additional_tip, additional_tip_percentage)) => {
+                        estimate.max_priority_fee_per_gas += max_additional_tip
+                            .min(estimate.max_fee_per_gas * additional_tip_percentage);
+                        estimate.max_priority_fee_per_gas = estimate
+                            .max_priority_fee_per_gas
+                            .min(estimate.max_fee_per_gas);
```

Comment:
This logic is interesting. I see we are doing the same in `SubmitterGasPriceEstimator`, but doesn't this mean we take the additional tip from the buffer we have in the base fee estimate? Shouldn't we instead increase the max_fee_per_gas by that same amount we increase the tip?

Reply:
I think this is also a reasonable way to handle it. Updated.